AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/helper/ValidatorParameterHelper.java

```diff
@@ -138,72 +164,107 @@ protected final void printMessage(ExecutableElement element) {
+			index++;
+			if (index < getParamRequirements().size()) {
+				currentParameterRequirement = getParamRequirements().get(index);
+			} else {
+				currentParameterRequirement = null;
+			}
+		}
+
+		@Override
+		public void validate(ExecutableElement executableElement, IsValid valid) {
+			nextParameterRequirement();
+			for (VariableElement parameter : executableElement.getParameters()) {
+				if (!validate(parameter)) {
 					invalidate(executableElement, valid);
+					return;
 				}
+			}
 
-				++index;
+			for (ParameterRequirement expectedParameter : getParamRequirements()) {
+				if (expectedParameter.required() && !satisfiedParameterRequirements.contains(expectedParameter)) {
+					invalidate(executableElement, valid);
+					return;
+				}
+			}
+		}
+
+		private boolean validate(VariableElement parameter) {
+			if (currentParameterRequirement == null) {
+				return false;
+			}
+			if (currentParameterRequirement.isSatisfied(parameter)) {
+				satisfiedParameterRequirements.add(currentParameterRequirement);
+			} else if (!currentParameterRequirement.multiple()) {
+				nextParameterRequirement();
+			} else {
+				if (currentParameterRequirement.required() && !satisfiedParameterRequirements.contains(currentParameterRequirement)) {
+					return false;
+				} else {
+					nextParameterRequirement();
+					return validate(parameter);
+				}
 			}
+			return true;
 		}
 
+
 		@Override
-		public T anyType() {
-			return extendsType(CanonicalNameConstants.OBJECT);
+		protected String createMessage(ExecutableElement element) {
+			return super.createMessage(element) + " in the order above";
 		}
 	}
 
-	public class AnyOrderParameterValidator extends ParameterValidatorBase<ChainableParameterRequirement<AnyOrderParameterValidator>> {
+	public class AnyOrderParamValidator extends BaseParamValidator<AnyOrderParamValidator> {
+
+		private List<ParameterRequirement> satisfiedParameterRequirements = new ArrayList<ParameterRequirement>();
 
 		@Override
 		public void validate(ExecutableElement executableElement, IsValid valid) {
 			for (VariableElement parameter : executableElement.getParameters()) {
-				MethodParameter foundParameter = null;
+				ParameterRequirement foundParameter = null;
 
-				for (MethodParameter expectedParameter : expectedParameters) {
-					if (exactType(parameter, expectedParameter.typeName) || extendsType(parameter, expectedParameter.typeName)) {
+				for (ParameterRequirement expectedParameter : getParamRequirements()) {
+					if (expectedParameter.isSatisfied(parameter)) {
+						satisfiedParameterRequirements.add(expectedParameter);
 						foundParameter = expectedParameter;
 						break;
 					}
 				}
 
-				if (foundParameter != null) {
-					expectedParameters.remove(foundParameter);
-				} else {
+				if (foundParameter == null) {
 					invalidate(executableElement, valid);
 					return;
 				}
+
+				if (!foundParameter.multiple()) {
+					getParamRequirements().remove(foundParameter);
```

Comment:
@yDelouis This is not okay, because in the end you will loop throught the parameter requirements in `createMessage`, and not all will be displayed to the user. I will edit this to include an original parameter requirenments field as well.